### ISSUE TITLE
Indexer bolts use canonical URL 

### ISSUE THREAD

**Issue by jnioche** [DigitalPebble/storm-crawler#149]
We already have a parse filter that extracts canonical URLs. We should add to AbstractIndexerBolt the ability to use them instead of the URL when available

**Comment by jorgelbg** [comment]
Adding this in the `AbstractIndexerBolt` means that it should be then optional to use in the indexers? perhaps creating and additional method that returns the right URL based on the Tuple received by the bolt, something like:

```java
    protected String valueForURL(Tuple tuple) {
        String url = tuple.getStringByField("url");
        Metadata metadata = (Metadata) tuple.getValueByField("metadata");

        if (metadata.getFirstValue("canonical") != null) 
            url = metadata.getFirstValue("canonical");

        return url;
    }
```

should work but this must be used instead of the default `url` var extracted from the Tuple in the execute method. This means that we should change this in the existing indexer implementations (ES/Solr/StdOut). Perhaps something like:

```java
String url = valueForURL(tuple);
...
```

Any thoughts on this?

**Comment by jorgelbg** [comment]
Adding this in the `AbstractIndexerBolt` means that it should be then optional to use in the indexers? perhaps creating and additional method that returns the right URL based on the Tuple received by the bolt, something like:

```java
    protected String valueForURL(Tuple tuple) {
        String url = tuple.getStringByField("url");
        Metadata metadata = (Metadata) tuple.getValueByField("metadata");

        if (metadata.getFirstValue("canonical") != null) 
            url = metadata.getFirstValue("canonical");

        return url;
    }
```

should work but this must be used instead of the default `url` var extracted from the Tuple in the execute method. This means that we should change this in the existing indexer implementations (ES/Solr/StdOut). Perhaps something like:

```java
String url = valueForURL(tuple);
...
```

Any thoughts on this?

**Comment by jnioche** [comment]
That's pretty much what I had in mind. Later on we could do the same as in Nutch and for redirections decide which of the source or target URLs is the nicest to display. 

We'd probably need to check that the value of canonical is OK e.g. it belongs to the same domain and make its value absolute if necessary.

**Comment by jorgelbg** [comment]
I'm going to add the validation in the canonical URL and submit a PR with the change.

**Comment by jnioche** [comment]
Implemented in #161
